dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/implementers/enrollment/SetMandatoryFieldExecutorTest.java

```diff
@@ -0,0 +1,233 @@
+        secondProgramStage = createProgramStage( 'B', 0 );
+        secondProgramStage.setValidationStrategy( ValidationStrategy.ON_UPDATE_AND_INSERT );
+        dataElementB = createDataElement( 'B' );
+        ProgramStageDataElement programStageDataElementB = createProgramStageDataElement( secondProgramStage,
+            dataElementB, 0 );
+        secondProgramStage.setProgramStageDataElements( Sets.newHashSet( programStageDataElementB ) );
+
+        attribute = createTrackedEntityAttribute( 'A' );
+        attribute.setUid( ATTRIBUTE_ID );
+        attribute.setCode( ATTRIBUTE_CODE );
+
+        bundle = TrackerBundle.builder().build();
+        bundle.setPreheat( preheat );
+    }
+
+    @Test
+    void shouldReturnNoErrorWhenMandatoryFieldIsPresentForEnrollment()
+    {
+        when( preheat.getIdSchemes() ).thenReturn( TrackerIdSchemeParams.builder().build() );
+        when( preheat.getTrackedEntityAttribute( ATTRIBUTE_ID ) ).thenReturn( attribute );
+        bundle.setEnrollments( Lists.newArrayList( getEnrollmentWithMandatoryAttributeSet() ) );
+
+        Optional<ProgramRuleIssue> error = enrollmentValidatorToTest.executeEnrollmentRuleAction( bundle,
+            getEnrollmentWithMandatoryAttributeSet() );
+
+        assertFalse( error.isPresent() );
+    }
+
+    @Test
+    void shouldReturnNoErrorWhenMandatoryFieldIsPresentForEnrollmentUsingIdSchemeCode()
+    {
+        TrackerIdSchemeParams idSchemes = TrackerIdSchemeParams.builder()
+            .idScheme( TrackerIdSchemeParam.CODE )
+            .build();
+        when( preheat.getIdSchemes() ).thenReturn( idSchemes );
+        when( preheat.getTrackedEntityAttribute( ATTRIBUTE_ID ) ).thenReturn( attribute );
+        bundle.setEnrollments( Lists.newArrayList( getEnrollmentWithMandatoryAttributeSet( idSchemes ) ) );
+
+        Optional<ProgramRuleIssue> errors = enrollmentValidatorToTest.executeEnrollmentRuleAction( bundle,
+            getEnrollmentWithMandatoryAttributeSet( idSchemes ) );
+
+        assertFalse( errors.isPresent() );
+    }
+
+    @Test
+    void shouldReturnAnErrorWhenMandatoryFieldIsNotPresentForEnrollment()
+    {
+        when( preheat.getIdSchemes() ).thenReturn( TrackerIdSchemeParams.builder().build() );
+        when( preheat.getTrackedEntityAttribute( ATTRIBUTE_ID ) ).thenReturn( attribute );
+        bundle.setEnrollments( Lists.newArrayList( getEnrollmentWithMandatoryAttributeSet(),
+            getEnrollmentWithMandatoryAttributeNOTSet() ) );
+
+        Optional<ProgramRuleIssue> error = enrollmentValidatorToTest.executeEnrollmentRuleAction( bundle,
+            getEnrollmentWithMandatoryAttributeSet() );
+
+        assertFalse( error.isPresent() );
+
+        error = enrollmentValidatorToTest.executeEnrollmentRuleAction( bundle,
+            getEnrollmentWithMandatoryAttributeNOTSet() );
+
+        assertTrue( error.isPresent() );
+
+        error.ifPresent( e -> {
```

Comment:
Writing separate `assertEquals` for each field will lead to one assertion error and not execute subsequent assertions. We should avoid this pattern as it causes us a lot of pain in flaky integration tests. (Since this is a unit test it would not be so hard to fix, but lets use good patterns everywhere 😋)

I think you should be able to do

```java
assertEquals( new ProgramRuleIssue("RULE_ATTRIBUTE", ...), e.get());
```

(Maybe ProgramRuleIssue needs an `@EqualsAndHashcode` annotation)